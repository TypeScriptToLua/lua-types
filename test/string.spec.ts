import { describeForEachLuaTarget, tstl } from './test-utils';

describeForEachLuaTarget('string', (target) => {
    test('string.byte', () => {
        const lua = tstl(
            target,
            `
            const b = string.byte("d");
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('string.byte multiple', () => {
        const lua = tstl(
            target,
            `
            const [a, b, c, d] = string.byte("abcd", 1, 4);
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('string.char', () => {
        const lua = tstl(
            target,
            `
            const str = string.char(64, 65, 66);
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('string.dump', () => {
        const lua = tstl(
            target,
            `
            const str = string.dump(() => 5);
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('string.find', () => {
        const lua = tstl(
            target,
            `
            const result = string.find("abc", "b");
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('string.find destructure', () => {
        const lua = tstl(
            target,
            `
            const [start, end, matchString] = string.find("abc", "b");
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('string.find destructure', () => {
        const lua = tstl(
            target,
            `
            const [start, end, matchString, matchString2] = string.find("abc", "b", 5, false);
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('string.format', () => {
        const lua = tstl(
            target,
            `
            const str = string.format('%q', 'a string with "quotes" and \\n new line');
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('string.gmatch', () => {
        const lua = tstl(
            target,
            `
            const s = "hello world from Lua";
            for (const [w] of string.gmatch(s, "%a+")) {
                print(w)
            }
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('string.gmatch destructure', () => {
        const lua = tstl(
            target,
            `
            const s = "from=world, to=Lua";
            for (const [k, v] of string.gmatch(s, "(%w+)=(%w+)")) {
                print(k, v)
            }
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('string.gsub string', () => {
        const lua = tstl(
            target,
            `
            const result = string.gsub("hello world", "l", "R");
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('string.gsub string destructured', () => {
        const lua = tstl(
            target,
            `
            const [result, numOccurrences] = string.gsub("hello world", "l", "R");
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('string.gsub table', () => {
        const lua = tstl(
            target,
            `
            const result = string.gsub("hello world", "l", {l: "R"});
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('string.gsub function', () => {
        const lua = tstl(
            target,
            `
            const result = string.gsub("hello world", "l", (match: string) => "R");
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('string.match', () => {
        const lua = tstl(
            target,
            `
            const result = string.match("hello world", "l");
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('string.match destructured', () => {
        const lua = tstl(
            target,
            `
            const [match1, match2] = string.match("hello world", "l");
        `
        );

        expect(lua).toMatchSnapshot();
    });
});
