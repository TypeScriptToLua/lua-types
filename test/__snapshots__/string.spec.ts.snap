// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lua version 5.1 / string string.byte 1`] = `"b = string.byte(\\"d\\")"`;

exports[`Lua version 5.1 / string string.byte multiple 1`] = `"a, b, c, d = string.byte(\\"abcd\\", 1, 4)"`;

exports[`Lua version 5.1 / string string.char 1`] = `"str = string.char(64, 65, 66)"`;

exports[`Lua version 5.1 / string string.dump 1`] = `
"str = string.dump(
    function() return 5 end
)"
`;

exports[`Lua version 5.1 / string string.find 1`] = `
"result = {
    string.find(\\"abc\\", \\"b\\")
}"
`;

exports[`Lua version 5.1 / string string.find destructure 1`] = `"start, ____end, matchString = string.find(\\"abc\\", \\"b\\")"`;

exports[`Lua version 5.1 / string string.find destructure 2`] = `"start, ____end, matchString, matchString2 = string.find(\\"abc\\", \\"b\\", 5, false)"`;

exports[`Lua version 5.1 / string string.format 1`] = `"str = string.format(\\"%q\\", \\"a string with \\\\\\"quotes\\\\\\" and \\\\n new line\\")"`;

exports[`Lua version 5.1 / string string.gmatch 1`] = `
"s = \\"hello world from Lua\\"
for w in string.gmatch(s, \\"%a+\\") do
    print(w)
end"
`;

exports[`Lua version 5.1 / string string.gmatch destructure 1`] = `
"s = \\"from=world, to=Lua\\"
for k, v in string.gmatch(s, \\"(%w+)=(%w+)\\") do
    print(k, v)
end"
`;

exports[`Lua version 5.1 / string string.gsub function 1`] = `
"result = {
    string.gsub(
        \\"hello world\\",
        \\"l\\",
        function(match) return \\"R\\" end
    )
}"
`;

exports[`Lua version 5.1 / string string.gsub string 1`] = `
"result = {
    string.gsub(\\"hello world\\", \\"l\\", \\"R\\")
}"
`;

exports[`Lua version 5.1 / string string.gsub string destructured 1`] = `"result, numOccurrences = string.gsub(\\"hello world\\", \\"l\\", \\"R\\")"`;

exports[`Lua version 5.1 / string string.gsub table 1`] = `
"result = {
    string.gsub(\\"hello world\\", \\"l\\", {l = \\"R\\"})
}"
`;

exports[`Lua version 5.1 / string string.match 1`] = `
"result = {
    string.match(\\"hello world\\", \\"l\\")
}"
`;

exports[`Lua version 5.1 / string string.match destructured 1`] = `"match1, match2 = string.match(\\"hello world\\", \\"l\\")"`;

exports[`Lua version 5.1 / string string.rep 1`] = `"str = string.rep(\\"foo\\", 3)"`;

exports[`Lua version 5.2 / string string.byte 1`] = `"b = string.byte(\\"d\\")"`;

exports[`Lua version 5.2 / string string.byte multiple 1`] = `"a, b, c, d = string.byte(\\"abcd\\", 1, 4)"`;

exports[`Lua version 5.2 / string string.char 1`] = `"str = string.char(64, 65, 66)"`;

exports[`Lua version 5.2 / string string.dump 1`] = `
"str = string.dump(
    function() return 5 end
)"
`;

exports[`Lua version 5.2 / string string.find 1`] = `
"result = {
    string.find(\\"abc\\", \\"b\\")
}"
`;

exports[`Lua version 5.2 / string string.find destructure 1`] = `"start, ____end, matchString = string.find(\\"abc\\", \\"b\\")"`;

exports[`Lua version 5.2 / string string.find destructure 2`] = `"start, ____end, matchString, matchString2 = string.find(\\"abc\\", \\"b\\", 5, false)"`;

exports[`Lua version 5.2 / string string.format 1`] = `"str = string.format(\\"%q\\", \\"a string with \\\\\\"quotes\\\\\\" and \\\\n new line\\")"`;

exports[`Lua version 5.2 / string string.gmatch 1`] = `
"s = \\"hello world from Lua\\"
for w in string.gmatch(s, \\"%a+\\") do
    print(w)
end"
`;

exports[`Lua version 5.2 / string string.gmatch destructure 1`] = `
"s = \\"from=world, to=Lua\\"
for k, v in string.gmatch(s, \\"(%w+)=(%w+)\\") do
    print(k, v)
end"
`;

exports[`Lua version 5.2 / string string.gsub function 1`] = `
"result = {
    string.gsub(
        \\"hello world\\",
        \\"l\\",
        function(match) return \\"R\\" end
    )
}"
`;

exports[`Lua version 5.2 / string string.gsub string 1`] = `
"result = {
    string.gsub(\\"hello world\\", \\"l\\", \\"R\\")
}"
`;

exports[`Lua version 5.2 / string string.gsub string destructured 1`] = `"result, numOccurrences = string.gsub(\\"hello world\\", \\"l\\", \\"R\\")"`;

exports[`Lua version 5.2 / string string.gsub table 1`] = `
"result = {
    string.gsub(\\"hello world\\", \\"l\\", {l = \\"R\\"})
}"
`;

exports[`Lua version 5.2 / string string.match 1`] = `
"result = {
    string.match(\\"hello world\\", \\"l\\")
}"
`;

exports[`Lua version 5.2 / string string.match destructured 1`] = `"match1, match2 = string.match(\\"hello world\\", \\"l\\")"`;

exports[`Lua version 5.2 / string string.rep 1`] = `"str = string.rep(\\"foo\\", 3)"`;

exports[`Lua version 5.3 / string string.byte 1`] = `"b = string.byte(\\"d\\")"`;

exports[`Lua version 5.3 / string string.byte multiple 1`] = `"a, b, c, d = string.byte(\\"abcd\\", 1, 4)"`;

exports[`Lua version 5.3 / string string.char 1`] = `"str = string.char(64, 65, 66)"`;

exports[`Lua version 5.3 / string string.dump 1`] = `
"str = string.dump(
    function() return 5 end
)"
`;

exports[`Lua version 5.3 / string string.find 1`] = `
"result = {
    string.find(\\"abc\\", \\"b\\")
}"
`;

exports[`Lua version 5.3 / string string.find destructure 1`] = `"start, ____end, matchString = string.find(\\"abc\\", \\"b\\")"`;

exports[`Lua version 5.3 / string string.find destructure 2`] = `"start, ____end, matchString, matchString2 = string.find(\\"abc\\", \\"b\\", 5, false)"`;

exports[`Lua version 5.3 / string string.format 1`] = `"str = string.format(\\"%q\\", \\"a string with \\\\\\"quotes\\\\\\" and \\\\n new line\\")"`;

exports[`Lua version 5.3 / string string.gmatch 1`] = `
"s = \\"hello world from Lua\\"
for w in string.gmatch(s, \\"%a+\\") do
    print(w)
end"
`;

exports[`Lua version 5.3 / string string.gmatch destructure 1`] = `
"s = \\"from=world, to=Lua\\"
for k, v in string.gmatch(s, \\"(%w+)=(%w+)\\") do
    print(k, v)
end"
`;

exports[`Lua version 5.3 / string string.gsub function 1`] = `
"result = {
    string.gsub(
        \\"hello world\\",
        \\"l\\",
        function(match) return \\"R\\" end
    )
}"
`;

exports[`Lua version 5.3 / string string.gsub string 1`] = `
"result = {
    string.gsub(\\"hello world\\", \\"l\\", \\"R\\")
}"
`;

exports[`Lua version 5.3 / string string.gsub string destructured 1`] = `"result, numOccurrences = string.gsub(\\"hello world\\", \\"l\\", \\"R\\")"`;

exports[`Lua version 5.3 / string string.gsub table 1`] = `
"result = {
    string.gsub(\\"hello world\\", \\"l\\", {l = \\"R\\"})
}"
`;

exports[`Lua version 5.3 / string string.match 1`] = `
"result = {
    string.match(\\"hello world\\", \\"l\\")
}"
`;

exports[`Lua version 5.3 / string string.match destructured 1`] = `"match1, match2 = string.match(\\"hello world\\", \\"l\\")"`;

exports[`Lua version 5.3 / string string.rep 1`] = `"str = string.rep(\\"foo\\", 3)"`;

exports[`Lua version 5.4 / string string.byte 1`] = `"b = string.byte(\\"d\\")"`;

exports[`Lua version 5.4 / string string.byte multiple 1`] = `"a, b, c, d = string.byte(\\"abcd\\", 1, 4)"`;

exports[`Lua version 5.4 / string string.char 1`] = `"str = string.char(64, 65, 66)"`;

exports[`Lua version 5.4 / string string.dump 1`] = `
"str = string.dump(
    function() return 5 end
)"
`;

exports[`Lua version 5.4 / string string.find 1`] = `
"result = {
    string.find(\\"abc\\", \\"b\\")
}"
`;

exports[`Lua version 5.4 / string string.find destructure 1`] = `"start, ____end, matchString = string.find(\\"abc\\", \\"b\\")"`;

exports[`Lua version 5.4 / string string.find destructure 2`] = `"start, ____end, matchString, matchString2 = string.find(\\"abc\\", \\"b\\", 5, false)"`;

exports[`Lua version 5.4 / string string.format 1`] = `"str = string.format(\\"%q\\", \\"a string with \\\\\\"quotes\\\\\\" and \\\\n new line\\")"`;

exports[`Lua version 5.4 / string string.gmatch 1`] = `
"s = \\"hello world from Lua\\"
for w in string.gmatch(s, \\"%a+\\") do
    print(w)
end"
`;

exports[`Lua version 5.4 / string string.gmatch destructure 1`] = `
"s = \\"from=world, to=Lua\\"
for k, v in string.gmatch(s, \\"(%w+)=(%w+)\\") do
    print(k, v)
end"
`;

exports[`Lua version 5.4 / string string.gsub function 1`] = `
"result = {
    string.gsub(
        \\"hello world\\",
        \\"l\\",
        function(match) return \\"R\\" end
    )
}"
`;

exports[`Lua version 5.4 / string string.gsub string 1`] = `
"result = {
    string.gsub(\\"hello world\\", \\"l\\", \\"R\\")
}"
`;

exports[`Lua version 5.4 / string string.gsub string destructured 1`] = `"result, numOccurrences = string.gsub(\\"hello world\\", \\"l\\", \\"R\\")"`;

exports[`Lua version 5.4 / string string.gsub table 1`] = `
"result = {
    string.gsub(\\"hello world\\", \\"l\\", {l = \\"R\\"})
}"
`;

exports[`Lua version 5.4 / string string.match 1`] = `
"result = {
    string.match(\\"hello world\\", \\"l\\")
}"
`;

exports[`Lua version 5.4 / string string.match destructured 1`] = `"match1, match2 = string.match(\\"hello world\\", \\"l\\")"`;

exports[`Lua version 5.4 / string string.rep 1`] = `"str = string.rep(\\"foo\\", 3)"`;

exports[`Lua version JIT / string string.byte 1`] = `"b = string.byte(\\"d\\")"`;

exports[`Lua version JIT / string string.byte multiple 1`] = `"a, b, c, d = string.byte(\\"abcd\\", 1, 4)"`;

exports[`Lua version JIT / string string.char 1`] = `"str = string.char(64, 65, 66)"`;

exports[`Lua version JIT / string string.dump 1`] = `
"str = string.dump(
    function() return 5 end
)"
`;

exports[`Lua version JIT / string string.find 1`] = `
"result = {
    string.find(\\"abc\\", \\"b\\")
}"
`;

exports[`Lua version JIT / string string.find destructure 1`] = `"start, ____end, matchString = string.find(\\"abc\\", \\"b\\")"`;

exports[`Lua version JIT / string string.find destructure 2`] = `"start, ____end, matchString, matchString2 = string.find(\\"abc\\", \\"b\\", 5, false)"`;

exports[`Lua version JIT / string string.format 1`] = `"str = string.format(\\"%q\\", \\"a string with \\\\\\"quotes\\\\\\" and \\\\n new line\\")"`;

exports[`Lua version JIT / string string.gmatch 1`] = `
"s = \\"hello world from Lua\\"
for w in string.gmatch(s, \\"%a+\\") do
    print(w)
end"
`;

exports[`Lua version JIT / string string.gmatch destructure 1`] = `
"s = \\"from=world, to=Lua\\"
for k, v in string.gmatch(s, \\"(%w+)=(%w+)\\") do
    print(k, v)
end"
`;

exports[`Lua version JIT / string string.gsub function 1`] = `
"result = {
    string.gsub(
        \\"hello world\\",
        \\"l\\",
        function(match) return \\"R\\" end
    )
}"
`;

exports[`Lua version JIT / string string.gsub string 1`] = `
"result = {
    string.gsub(\\"hello world\\", \\"l\\", \\"R\\")
}"
`;

exports[`Lua version JIT / string string.gsub string destructured 1`] = `"result, numOccurrences = string.gsub(\\"hello world\\", \\"l\\", \\"R\\")"`;

exports[`Lua version JIT / string string.gsub table 1`] = `
"result = {
    string.gsub(\\"hello world\\", \\"l\\", {l = \\"R\\"})
}"
`;

exports[`Lua version JIT / string string.match 1`] = `
"result = {
    string.match(\\"hello world\\", \\"l\\")
}"
`;

exports[`Lua version JIT / string string.match destructured 1`] = `"match1, match2 = string.match(\\"hello world\\", \\"l\\")"`;

exports[`Lua version JIT / string string.rep 1`] = `"str = string.rep(\\"foo\\", 3)"`;
